### PR TITLE
Extend the functionality of allow/block/reject commands to work with rules

### PR DESCRIPTION
**Pull request by Cropi**
These commands are now able to work with either a **device id** or a specific **device rule**.

### CODE REVIEW

File: src/CLI/usbguard-allow-device.cpp [hunk 1]
```diff
@@ -75,14 +76,47 @@ namespace usbguard
...
+      }
+      catch (...) {
+        showHelp(std::cerr);
+        return EXIT_FAILURE;
+      }
+
+      /*
+       * Create a ruleset which will contain the requested rule
+      */
+      auto my_rule_set = std::dynamic_pointer_cast<RuleSet>(std::make_shared<MemoryRuleSet>(nullptr));
```
Comment (dkopecek) [inline src/CLI/usbguard-allow-device.cpp]: I don't think you have to create a RuleSet here. Wouldn't it be possible to use the [Rule::appliesTo](https://github.com/USBGuard/usbguard/blob/27801e8800b6c2ebf34a6104d318ae541ff66706/src/Library/public/usbguard/Rule.hpp#L490) method? i.e. something like:
```
if (rule.appliesTo(rule_device)) ... 
```

File: src/CLI/usbguard-allow-device.cpp [hunk 2]
```diff
@@ -22,6 +22,7 @@
 
 #include "usbguard.hpp"
 #include "usbguard-allow-device.hpp"
+#include "PolicyGenerator.hpp"
```
Comment (dkopecek) [inline src/CLI/usbguard-allow-device.cpp]: Including this is unnecessary, I think you want `#include "usbguard/Rule.hpp"`.

File: src/CLI/usbguard-allow-device.cpp [hunk 3]
```diff
@@ -75,14 +76,47 @@ namespace usbguard
...
-    if (argc != 1) {
-      showHelp(std::cerr);
-      return EXIT_FAILURE;
+    usbguard::IPCClient ipc(/*connected=*/true);
+
+    if (argc == 1) {
+      id = std::stoul(argv[0]);
+      ipc.applyDevicePolicy(id, Rule::Target::Allow, permanent);
+    }
+    else {
```
Comment (dkopecek) [inline src/CLI/usbguard-allow-device.cpp]: Does this work with `argc==0`?

File: src/CLI/usbguard-block-device.cpp [hunk 4]
```diff
@@ -75,14 +76,47 @@ namespace usbguard
...
+      for (int i = 1; i < argc; i++) {
+        rule_string.append(" ").append(argv[i]);
+      }
+
+      usbguard::Rule rule;
+      try {
+        rule = Rule::fromString(rule_string);
+      }
+      catch (...) {
+        showHelp(std::cerr);
```
[the remark this answers is not in the document]
Comment (Cropi) [reply src/CLI/usbguard-block-device.cpp]: @radosroka  Is it enough to mention that a RULE_EXCEPTION happened or should it be more specific?

File: src/CLI/usbguard-allow-device.cpp (hunk 1 above)
Comment (Cropi) [reply src/CLI/usbguard-allow-device.cpp]: Yes, you are right. It is a lot more elegant solution.

File: src/CLI/usbguard-allow-device.cpp [hunk 5]
```diff
@@ -75,14 +76,47 @@ namespace usbguard
...
+        return EXIT_FAILURE;
+      }
+
+      /*
+       * Create a ruleset which will contain the requested rule
+      */
+      auto my_rule_set = std::dynamic_pointer_cast<RuleSet>(std::make_shared<MemoryRuleSet>(nullptr));
+      my_rule_set->appendRule(rule);
+
+      for (auto rule_device : ipc.listDevices("match")) {
```
Comment (Cropi) [inline src/CLI/usbguard-allow-device.cpp]: I am not sure if the **allow|block|reject-device** commands should affect all the devices or just a few of them, e.g. those which are "allowed", "blocked".
My question is whether I should replace
`for (auto rule_device : ipc.listDevices("match")) {`
with
`for (auto rule_device : ipc.listDevices(argv[0])) {`
@dkopecek

File: src/CLI/usbguard-block-device.cpp (hunk 4 above)
[the remark this answers is not in the document]
Comment (radosroka) [reply src/CLI/usbguard-block-device.cpp]: It is OK now.

File: src/CLI/usbguard-allow-device.cpp [hunk 6]
```diff
@@ -75,14 +76,47 @@ namespace usbguard
...
+      for (int i = 1; i < argc; i++) {
+        rule_string.append(" ").append(argv[i]);
+      }
+
+      usbguard::Rule rule;
+      try {
+        rule = Rule::fromString(rule_string);
+      }
+      catch (...) {
+        showHelp(std::cerr);
```
[the remark this answers is not in the document]
Comment (radosroka) [reply src/CLI/usbguard-allow-device.cpp]: Done.

File: src/CLI/usbguard-allow-device.cpp (hunk 5 above)
Comment (dkopecek) [reply src/CLI/usbguard-allow-device.cpp]: I'm not sure what listDevices(argv[0]) means but I think you can use the match rule constructed from the command line argument directly as a query in listDevices(). But listing all devices and applying the rule only on the matching ones is equivalent. The difference is only which side (cli or daemon) is doing the matching.

Comment (Cropi) [reply src/CLI/usbguard-allow-device.cpp]: Sorry, my question was not well formed. 
I choose the first option - the CLI is doing the matching.

Comment (dkopecek) [reply src/CLI/usbguard-allow-device.cpp]: Ok

**Comment by Cropi** [comment]
All the requested changes are finished now and is ready for merge.

File: .travis.yml [hunk 7]
```diff
@@ -73,7 +73,7 @@ before_install:
 - sudo apt-get install -y asciidoc
 - sudo apt-get install -y umockdev libumockdev-dev
 - sudo gem install coveralls-lcov
-- curl -JLo /tmp/astyle.tgz --remote-header-name 'https://netix.dl.sourceforge.net/project/astyle/astyle/astyle%203.1/astyle_3.1_linux.tar.gz'
+- curl -JLo /tmp/astyle.tgz --remote-header-name 'netcologne.dl.sourceforge.net/project/astyle/astyle/astyle%203.1/astyle_3.1_linux.tar.gz'
```
Comment (radosroka) [inline .travis.yml]: forgot https://?

**Comment by tammar96** [comment]
Can you please enhance the documentation? An example would be great.

File: src/Common/Utility.hpp [hunk 8]
```diff
@@ -294,6 +294,24 @@ namespace usbguard
...
   };
 
+
+  template<typename T>
+  std::string joinElements(T begin, T end, const std::string& separator = " ")
+  {
+    std::ostringstream ss;
+
+    if (begin != end) {
+      ss << *begin++;
```
Comment (radosroka) [inline src/Common/Utility.hpp]: I think that *(begin++) is much  more readable.

Comment (Cropi) [reply src/Common/Utility.hpp]: Okay

**Comment by Cropi** [comment]
> I tried to test this pull request but I was not able to comprehend what do you want to achieve by the documentation. Can you please enhance the documentation with an usage example?

Some examples related to allow-device have been added.

**Comment by tammar96** [comment]
I tried it and it looks like that it is working flawlessly. :+1:

File: src/Common/Utility.hpp (hunk 8 above)
Comment (tammar96) [reply src/Common/Utility.hpp]: I agree with @radosroka

**Comment by radosroka** [comment]
Great Thanks.

**Pull request merged**